text/XXX-template-poly.md

```diff
@@ -0,0 +1,154 @@
+- Title: Template poly redesign using sort poly
+
+- Drivers: Gaëtan Gilbert
+
+----
+
+# Summary
+
+Template poly currently provides 2 features in 1: implicit universe
+instances and adhoc sort polymorphism handling only Prop/Type.
+
+We want to make it less adhoc and base it on more well understood
+systems such as regular sort poly.
+
+This should also allow to stop producing useless constraints (like u <= prod.u0)
+when a template inductive is fully applied.
+
+# Detailed design
+
+## Declaration
+
+An inductive may be declared to use implicit univ instances when:
+
+- it is non mutual (nested is ok) (this rule is in master but maybe we will remove it)
+- its univ declaration has only univ variables unbounded from below
+- its univ declaration has at most 1 qvar variable.
+  If it has one, it is the quality of the output sort.
+- each parameter type either does not mention the bound univs (CLOSED),
+  or is of the form `forall args, Type@{q|u}` (BINDING)
+  where the `args` types do not mention the bound univs and qvar,
+  `q` is either Type or the unique bound qvar,
+  and `u` is either constant or one of the bound univ levels.
+- the indices types and constructor types do not mention the bound univs and qvar.
+
+NB: the qvar (if there is one) must be "above prop" so should not appear in relevance marks
+(it's always `Relevant`) so eg
```

Comment:
not fully clear if we want to keep the above_prop restriction
it makes this relevance issue easier but at the cost of needing the kernel to understand above_prop

Reply:
In fact we may not need to understand above_prop, we can allow template qualities to be instantiated by anything. 
However taking the max of the argument qualities to get the implicit instance may then fail.

This means for instance we can have `sigma@{q|a b} : forall (A:Type@{q|a}) (B:A -> Type@{q|b}), Type@{q|max(a,b)}` such that
`sigma True (fun _ => nat) : Set`
`sigma True (fun _ => True) : Prop`
`sigma sTrue (fun _ => sTrue) : SProp`
`sigma A B : Type@{q|...}` if A and B are in the same qvar `q`

and `sigma True (fun _ => sTrue)` and `sigma A B` with A and B at different qvars are errors.